src/string_id_null_ids.cpp

```diff
@@ -1,34 +1,41 @@
 #include "string_id.h"
 
-// These are centralized in this file because they must appear in one and only one compilation unit,
-// and it's problematic to define them in the respective cpp files for each class.
-// Very repitious, so define them with a macro.
-#define MAKE_NULL_ID( type, ... ) \
-class type; \
-template<> const string_id<type> &string_id<type>::NULL_ID() { \
-    static string_id<type> id = string_id<type>( __VA_ARGS__ ); \
-    return id; \
-}
+#define MAKE_NULL_ID( type, ... )  class type;  template<> const string_id<type> &string_id<type>::NULL_ID() {  static string_id<type> id = string_id<type>( __VA_ARGS__ );  return id; }
```

Comment:
I don't have the slighliest idea of how anything in this .cpp works. I barely know enough about preprocessors to make guards and include whatever header I need.
But after a week of not being able to compile, I just tried by whatever means to make it work, hence copypasting things I don't understand and asking in IRC for help. 
And to remark by total lack of knowledge and experience I just applied another fix by @Graywolfe813.